modules/transformer-kinesis/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/transformer/kinesis/parquet/ParquetSink.scala

```diff
@@ -0,0 +1,86 @@
+package com.snowplowanalytics.snowplow.rdbloader.transformer.kinesis.parquet
+
+import cats.Functor
+import cats.data.EitherT
+import cats.effect.{Blocker, Concurrent, ContextShift, Timer}
+import cats.implicits._
+import com.github.mjakubowski84.parquet4s.RowParquetRecord
+import com.github.mjakubowski84.parquet4s.parquet.viaParquet
+import com.snowplowanalytics.snowplow.badrows.FailureDetails
+import com.snowplowanalytics.snowplow.rdbloader.common.LoaderMessage.TypesInfo.WideRow
+import com.snowplowanalytics.snowplow.rdbloader.common.transformation.Transformed
+import com.snowplowanalytics.snowplow.rdbloader.common.transformation.Transformed.Data.ParquetData.FieldWithValue
+import com.snowplowanalytics.snowplow.rdbloader.common.transformation.parquet.fields.AllFields
+import com.snowplowanalytics.snowplow.rdbloader.common.transformation.parquet.{AtomicFieldsProvider, NonAtomicFieldsProvider}
+import com.snowplowanalytics.snowplow.rdbloader.transformer.kinesis.parquet.Codecs._
+import com.snowplowanalytics.snowplow.rdbloader.transformer.kinesis.sinks.{SinkPath, TransformedDataOps, Window}
+import com.snowplowanalytics.snowplow.rdbloader.transformer.kinesis.{Resources, State}
+import fs2.{Pipe, Stream}
+import org.apache.parquet.schema.MessageType
+
+import java.net.URI
+
+object ParquetSink {
+
+  def parquetSink[F[_] : Concurrent : ContextShift : Timer](resources: Resources[F],
+                                                            uri: URI)
+                                                           (window: Window)
+                                                           (path: SinkPath): Pipe[F, Transformed.Data, Unit] = {
+    transformedData =>
+
+      val targetPath = uri.toString + window.getDir + "/" + path.value
+      val schemaCreation = createSchemaFromTypes(resources, window).value
+
+      Stream.eval(schemaCreation).flatMap {
+        case Left(error) =>
+          println(error) //TODO
```

Comment:
A difficult problem!  What are your thoughts?

I suppose `parquetSink` could return `EitherT[F, FailureDetails, Pipe[F, Transformed.Data, Unit]]` but then you need to change a lot of upstream code.

Or you need to change upstream code so that the schemas or fields are passed down into this `parquetSink` function.  It should be possible because we know all the schemas already.

On the other hand, this error is unexpected, because we already know the schemas exist in Iglu.  So you could `Stream.raiseError(new RuntimeException("..."))` which would let the app checkpoint, exit, and restart.

Reply:
We have to create schema when the sink is actually used (micro batch flush or window ends), cause only then we have all we need. As you said this error is very unlikely to happen, so I would go now with `Stream.raiseError`